metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/adapters/maxmind/GeoLiteDatabase.java

```diff
@@ -0,0 +1,165 @@
+      } finally {
+        unlockIfNecessary();
+      }
+    } else if (hdfsFile.endsWith(GeoLiteDatabase.EXTENSION_TAR_GZ)) {
+      lockIfNecessary();
+      try (TarArchiveInputStream is = new TarArchiveInputStream(
+          new GZIPInputStream(fs.open(new Path(hdfsFile))))) {
+        // Need to find the mmdb entry.
+        TarArchiveEntry entry = is.getNextTarEntry();
+        while (entry != null) {
+          if (entry.isFile() && entry.getName().endsWith(GeoLiteDatabase.EXTENSION_MMDB)) {
+            try(InputStream mmdb = new BufferedInputStream(is))
+            { // Read directly from tarInput
+              setReader(readNewDatabase(getReader(), hdfsFile, mmdb));
+              break; // Don't care about the other entries, leave immediately
+            }
+          }
+          entry = is.getNextTarEntry();
+        }
+      } catch (IOException e) {
+        handleDatabaseIOException(hdfsFile, e);
+      } finally {
+        unlockIfNecessary();
+      }
+    }
+  }
+
+  default void handleDatabaseIOException(String hdfsFile, IOException e) {
+    LOG.error("[Metron] Unable to open new database file {}", hdfsFile, e);
+    throw new IllegalStateException("[Metron] Unable to update MaxMind database");
+  }
+
+  default DatabaseReader readNewDatabase(DatabaseReader reader, String hdfsFile, InputStream is) throws IOException {
+    LOG.info("[Metron] Update to GeoIP data started with {}", hdfsFile);
+    // InputStream based DatabaseReaders are always in memory.
+    DatabaseReader newReader = new DatabaseReader.Builder(is).withCache(new CHMCache()).build();
+    // If we've never set a reader, don't close the old one
+    if (reader != null) {
+      reader.close();
+    }
+    LOG.info("[Metron] Finished update to GeoIP data started with {}", hdfsFile);
+    return newReader;
+  }
+
+  default FileSystem getFileSystem() {
+    FileSystem fs;
+    try {
+      fs = FileSystem.get(new Configuration());
+    } catch (IOException e) {
+      LOG.error("[Metron] Unable to retrieve get HDFS FileSystem");
+      throw new IllegalStateException("[Metron] Unable to get HDFS FileSystem");
+    }
+    return fs;
+  }
+
+  default String convertNullToEmptyString(Object raw) {
+    return raw == null ? "" : String.valueOf(raw);
+  }
+
+  default boolean invalidIp(String ip) {
+    LOG.trace("[Metron] Called validateIp({})", ip);
+    InetAddress addr;
+    try {
+      addr = InetAddress.getByName(ip);
+    } catch (UnknownHostException e) {
+      LOG.warn("[Metron] No result found for IP {}", ip, e);
+      return true;
+    }
+    if (isIneligibleAddress(ip, addr)) {
+      LOG.debug("[Metron] IP ineligible for lookup {}", ip);
+      return true;
+    }
+    return false;
+  }
+
+  default boolean isIneligibleAddress(String ipStr, InetAddress addr) {
+    return addr.isAnyLocalAddress() || addr.isLoopbackAddress()
+        || addr.isSiteLocalAddress() || addr.isMulticastAddress()
+        || !ipvalidator.isValidInet4Address(ipStr);
```

Comment:
Does the Maxmind API say specifically to exclude all of these?  Like for example IPv6?  Of course, these all seem like logical exclusions, but I'd rather just do a lookup in the database and have it tell me its not there versus making assumptions about what is not there.

Reply:
It doesn't.

Re: IPv6, I believe this is holdover from the earlier DBs and we haven't done any testing or validation. In addition, IPv6 throughout our system is problematic and would require further changes, so I didn't touch anything here.

Re: other clauses. This was done to limit calls (and by association readlocks) to the DB for items that wouldn't end up with results anyway.  If we want to reevaluate them, I'm definitely open.

Reply:
We can leave as-is.  No sense in trying to tackle this now.